File: week_2/minji/BOJ_9663.py
'''
pypy 통과

퀸의 경우 같은 열이나 행에 있으면 공격 가능
-> 퀸은 한 행에 하나씩 존재해야한다. 그래서 1차원 배열을 사용해서 같은 행에 존재하는 경우 배제

boards 의미
-> board[x]=y-> x+1번째 행 y+1번째 열에 퀸이 놓임
'''
n=int(input())
boards=[0]*n
ans=0
def check(x):
    for i in range(x) :
        if boards[x]==boards[i] :
            return False
        elif abs(boards[x]-boards[i])==abs : # 같은 열에 존재하는지 확인
            return False
    return True
def dfs(x):
    global ans
    if x==n :
        ans+=1
        return
    else :
        for i in range(n) :
            boards[x]=i
            if check(x)==True:
                dfs(x+1)

dfs(0)
print(ans)

File: week_02/minji/BOJ_9663.java
package minji;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class BOJ_9663 {
    static int n;
    static int ans;
    static List<Integer> boards=new ArrayList<>();

    public static boolean check(int x){
        for(int i=0; i<x; i++){
            if(boards.get(x) == boards.get(i)){
                return false;
            }else if(Math.abs(boards.get(x) - boards.get(i))==x-i){
                return false;
            }
        }
        return true;
    }
    public static void dfs(int x){
        if(x==n){
            ans+=1;
            return;
        }
        for(int i=0; i<n; i++){
            boards.set(x, i);
            if(check(x)==true){
                dfs(x+1);
            }
        }
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        for(int i=0; i<n; i++){
            boards.add(0);
        }
        ans=0;
        dfs(0);
        System.out.println(ans);
    }
}
